### PR TITLE
Add support for harvesting records with VuFind fields with OAI-PMH.

### PR DESCRIPTION
**Pull request by EreMaijala**
This is a proposal to add an oai_vufind format to the OAI-PMH provider. This would allow harvesting of diverse metadata records with more granular metadata than the oai_dc format can provide.

TODO:
- [x] Decide whether having the oai_dc record as the base makes sense. I did it so that there's at least some usable XML instead of just a chunk of JSON. [yes]
- [x] Should we try to come up with an XML translation of the JSON record? It won't be easy to make it nice since we have field contents as keys. One option would be to introduce new formats for those cases. [no]
- [x] Is the prefix and namespace/xsd ok? Adding a real XML schema file for the current implementation would be straightforward since it's oai_dc with one additional field in another namespace. [yes]

### CODE REVIEW

File: module/VuFindApi/config/module.config.php [hunk 1]
```diff
@@ -12,6 +12,12 @@
             'SearchApi' => 'VuFindApi\Controller\SearchApiController',
         ],
     ],
+    'service_manager' => [
+        'allow_override' => true,
```
Comment (demiankatz) [inline module/VuFindApi/config/module.config.php]: Do you need this allow_override setting here? Won't it be inherited from the main VuFind module?

File: module/VuFindApi/src/VuFindApi/Controller/Factory.php [hunk 2]
```diff
@@ -67,11 +66,10 @@ public static function getApiController(ServiceManager $sm)
      */
     public static function getSearchApiController(ServiceManager $sm)
     {
-        $recordFields = $sm->get('VuFind\Config\YamlReader')
-            ->get('SearchApiRecordFields.yaml');
-        $helperManager = $sm->get('ViewHelperManager');
-        $rf = new RecordFormatter($recordFields, $helperManager);
-        $controller = new SearchApiController($sm, $rf, new FacetFormatter());
-        return $controller;
+        return new SearchApiController(
```
Comment (demiankatz) [inline module/VuFindApi/src/VuFindApi/Controller/Factory.php]: This refactoring of the RecordFormatter to a service seems like it might make sense as a separate commit, since it improves extensibility and has nothing to do with the rest of this PR. Is there any reason I shouldn't extract this part of the change and commit it to master?

File: module/VuFind/src/VuFind/OAI/Server.php [hunk 3]
```diff
@@ -497,6 +540,16 @@ protected function initializeSettings(\Zend\Config\Config $config)
         if (isset($config->OAI->set_query)) {
             $this->setQueries = $config->OAI->set_query->toArray();
         }
+
+        if (!empty($config->OAI->vufind_api_format_fields)) {
+            $this->vufindApiFields = explode(
+                ',', $config->OAI->vufind_api_format_fields
+            );
+            $this->metadataFormats['oai_vufind'] = [
+                'schema' => 'http://vufind.org/oai_vufind-6.0.xsd',
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/OAI/Server.php]: Is the 6.0 here because you're assuming this will be released as part of VuFind 6.0? Probably a reasonable assumption due to the backward-breaking constructor signature change. Just making sure I understand what you're doing!

If you do want to create an xsd document to serve at this URL, I could certainly make that happen... though all things being equal, it might be better to use a directory, like `https://vufind.org/xsd/oai_vufind-6.0.xsd`.

**Review by demiankatz — COMMENTED** [review]
Thanks, @EreMaijala. Here are some comments and questions:

- Assuming there is no obvious use case for XML translation, I'd certainly be happy with what you're doing here as the simplest possible implementation.
- Regarding the JSON vs. XML translation issue, should we call this format oai_vufind_json so that we can address the XML version more seamlessly if it turns out there's a use case for that? Even if that never happens, there's no harm in a descriptive name.

See also individual comments below.

File: module/VuFindApi/src/VuFindApi/Controller/Factory.php (hunk 2 above)
Comment (EreMaijala) [reply module/VuFindApi/src/VuFindApi/Controller/Factory.php]: That'd be great, but then the same should probably be done with FacetFormatter.

File: module/VuFind/src/VuFind/OAI/Server.php (hunk 3 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAI/Server.php]: Yes, I'm hoping this would be part of 6.0. And those paths were just a suggestion. I'll see if I can manage to create a suitable xsd.

File: module/VuFindApi/config/module.config.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFindApi/config/module.config.php]: I wasn't sure about that, but in any case if VuFindApi somehow became less dependent on the main module in the future, I think it wouldn't hurt to be explicit here. Do you think it would be better without?

**Comment by EreMaijala** [comment]
oai_vufind_json makes sense, I'll change that and the other bits.

File: module/VuFind/src/VuFind/OAI/Server.php (hunk 3 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/OAI/Server.php]: Then again, maybe the format could be 1.0 too. Not sure which one is better.

**Comment by demiankatz** [comment]
@EreMaijala, as discussed, I have committed the refactoring of API formatters to services to master in order to simplify this PR.

**Comment by EreMaijala** [comment]
The metadata prefix is now oai_vufind_json and schema version 1.0. I changed the structure of the response to distinquish the DC part from the JSON part better.

Here's what I managed to whip up with my limited XSD skills:
```
<xs:schema targetNamespace="http://vufind.org/oai_vufind_json-1.0"
    xmlns:oai_vufind_json="http://vufind.org/oai_vufind_json-1.0"
    xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    elementFormDefault="qualified" attributeFormDefault="unqualified">

    <xs:import namespace="http://www.openarchives.org/OAI/2.0/oai_dc/"
        schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc.xsd"/>

    <xs:element name="record" type="oai_vufind_json:recordType"/>

    <xs:complexType name="recordType">
        <xs:sequence>
            <xs:element ref="oai_dc:dc"/>
            <xs:element name="metadata" type="oai_vufind_json:metadataType"/>
        </xs:sequence>
    </xs:complexType>

    <xs:complexType name="metadataType" mixed="true">
        <xs:attribute name="type" type="xs:string" use="required"/>
    </xs:complexType>
</xs:schema>


```

**Comment by demiankatz** [comment]
@EreMaijala, my XSD skills are even more limited than yours, as I've never had much cause to work with them. At a glance, though, it looks fine. Have you tested whether things validate as expected? In case it helps, I've actually posted the document at `https://vufind.org/xsd/oai_vufind_json-1.0.xsd`. Let me know if any changes are needed.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala. Since the XSD is online now, it should be possible to test validation with online XSD at this point too.

**Comment by EreMaijala** [comment]
I tried to validate a full response with several validators, but I'm just getting weird errors that are not very useful. While the XML and the XSD files validate fine by themselves, either the validator is unable to read an XSD or complains that "White Spaces Are Required Between PublicId And SystemId., Line '1', Column '50'."

**Comment by EreMaijala** [comment]
@demiankatz Thanks, but that didn't help. But I finally realized that the validator didn't follow http->https redirect. I switched the schema location to https and now it validates. :)

**Comment by demiankatz** [comment]
@EreMaijala, another thought on this: since the record formatter is only used for a fairly specialized edge case, maybe we should not inject it as a constructor parameter, but instead add a setRecordFormatter method. That would keep the constructor signature backward-compatible, enabling us to merge this into 5.1 as a new feature rather than having to wait for 6.0. We could just set things up so that the new OAI format is disabled unless the record formatter is set.

On a semi-related note, are there possible negative consequences for using this format with the authority index? I can't remember if there are biblio-specific assumptions being made.

Anyway, just a thought -- let me know if you'd like me to test anything or help with making any changes. I'm also happy to hold off on more progress here until the actual output format has been discussed further, but if you're happy with it, I'd love to get this wrapped up so we can reconcile it with your other OAI-PMH changes.

**Comment by EreMaijala** [comment]
@demiankatz Done, sorry for delay. I had to change the flow a bit to defer the initialization of metadata formats to when the actual request is served to make sure we have the record formatter and config setup.

There should be no issue with using the format with authority records, though there won't be many useful fields with the default field config.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala. Two notes:

1.) We'll need to revert the test changes to get that passing again. I'm happy to take care of that -- easy enough!

2.) What do you think about still calling initializeMetadataFormats in the constructor, but also calling it at the bottom of setRecordFormatter? That seems perhaps a little more efficient than calling it on every request and ensures that at least some formats are set even if no request has been made yet.

**Comment by EreMaijala** [comment]
Tests should be fixed now. And there's normally just a single request served per invokation, so getResponse won't called multiple times, right?

**Comment by demiankatz** [comment]
Fair point. Scratch efficiency from my proposal. Just feels less side-effect-driven that way... but not a big deal.

**Comment by EreMaijala** [comment]
I'll revise. I'd still call it after initializeSettings() has been called unlike before. Feels like more obvious order to have settings first.

**Comment by demiankatz** [comment]
@EreMaijala, I've done some more testing and found and fixed a couple of minor issues:

1.) The ListMetadataFormats action never included the oai_vufind_json option; I added a special case to make that work right, and refactored code so we now have a standard supportsVuFindMetadata() method for checking this special case. I also refactored the code to build the VuFind metadata to a support method, which makes everything a bit more readable.

2.) There was a bug where the initialization of vufindApiFields was defaulting to an array containing an empty string, which evaluates to true. I added an array_filter to strip that out.

I also revised the initializeMetadataFormats logic so it is now called on demand, instead of running twice on every request. Hopefully that's the cleanest possible approach.

Please take one last look and let me know if you have any concerns. If not, I think I can check all the boxes above and merge this as soon as I hear from you.

**Comment by EreMaijala** [comment]
@demiankatz Thanks for the fixes and improvements, they all look good to me!

**Comment by demiankatz** [comment]
Thanks -- merged!
